ncm-opennebula/src/main/perl/OpenNebula/AII.pm

```diff
@@ -0,0 +1,433 @@
+
+=cut
+
+sub aii_configure
+{
+    my ($self, $config, $path) = @_;
+    my $tree = $config->getElement($path)->getTree();
+    my $createimage = $tree->{image};
+    $self->info("Create VM image flag is set to: $createimage");
+    my $createvmtemplate = $tree->{template};
+    $self->info("Create VM template flag is set to: $createvmtemplate");
+    my $permissions = $self->get_permissions($config);
+
+    my $fqdn = $self->get_fqdn($config);
+
+    # Set one endpoint RPC connector
+    my $one = $self->read_one_aii_conf($config);
+    if (!$one) {
+        $self->error("No ONE instance returned");
+        return 0;
+    }
+
+    # Check RPC endpoint and OpenNebula version
+    my $oneversion = $self->is_supported_one_version($one);
+    return 0 if !$oneversion;
+
+    my %images = $self->get_images($config);
+    $self->remove_or_create_vm_images($one, $createimage, \%images, $permissions);
+
+    my %ars = $self->get_vnetars($config);
+    $self->remove_and_create_vn_ars($one, \%ars);
+
+    my $vmtemplatetxt = $self->get_vmtemplate($config, $oneversion);
+    my $vmtemplate = $self->remove_or_create_vm_template($one, $fqdn, $createvmtemplate, $vmtemplatetxt, $permissions);
+}
+
+=item aii_install
+
+Based on Quattor template this method:
+
+Stops current running VM.
+Instantiates the new VM.
+
+=cut
+
+sub aii_install
+{
+    my ($self, $config, $path) = @_;
+    my (%opts, $vmtemplate);
+
+    my $tree = $config->getElement($path)->getTree();
+
+    my $instantiatevm = $tree->{vm};
+    $self->info("Instantiate VM flag is set to: $instantiatevm");
+    my $onhold = $tree->{onhold};
+    $self->info("Start VM onhold flag is set to: $onhold");
+    my $permissions = $self->get_permissions($config);
+
+    my $fqdn = $self->get_fqdn($config);
+
+    # Set one endpoint RPC connector
+    my $one = $self->read_one_aii_conf($config);
+    if (!$one) {
+        $self->error("No ONE instance returned");
+        return 0;
+    }
+
+    # Check RPC endpoint and OpenNebula version
+    my $oneversion = $self->is_supported_one_version($one);
+    return 0 if !$oneversion;
+
+    $self->stop_and_remove_one_vms($one, $fqdn);
+
+    my @vmtpl = $one->get_templates(qr{^$fqdn$});
+
+    if (@vmtpl) {
+        $vmtemplate = $vmtpl[0];
+        $self->info("Found VM template from ONE database: ", $vmtemplate->name);
+    } else {
+        $self->error("VM template is not available to instantiate VM: $fqdn");
+        return 0;
+    }
+
+    if ($instantiatevm) {
+        $self->debug(1, "Instantiate vm with name $fqdn with template ", $vmtemplate->name);
+
+        # Check that image is in READY state.
+        my @myimages = $one->get_images(qr{^${fqdn}\_vd[a-z]$});
+        $opts{max_iter} = $MAXITER;
+        foreach my $t (@myimages) {
+            # If something wrong happens set a timeout
+            my $imagestate = $t->wait_for_state("READY", %opts);
+
+            if ($imagestate) {
+                $self->info("VM Image status: READY ,OK");
```

Comment:
add at least `$fqdn` to message, maybe even template name?
also remove space before `,` to after `,`